src/otg/method/finemapping.py

```diff
@@ -0,0 +1,160 @@
+            )
+
+        elif outlier_method == "CARMA":
+            # unfinished work in progress
+            import carmapy.carmapy_c
+            from scipy import optimize
+
+            sumstats = pd.read_csv(
+                f"{target}_locus_sumstat_flip_check.txt.gz", sep="\t"
+            )
+            ld = pd.read_csv(f"{target}_locus_ukbb_ld.txt.gz", sep="\t", header=None)
+            outlier_tau = 0.04
+            index_list = sumstats.index.tolist()
+            z = sumstats["z"].tolist()
+            ld_matrix = np.asmatrix(ld)
+            modi_ld_S = ld_matrix
+
+            def ridge_fun(
+                x, modi_ld_S, index_list, temp_Sigma, z, outlier_tau, outlier_likelihood
+            ):
+                temp_ld_S = x * modi_ld_S + (1 - x) * np.eye(modi_ld_S.shape[0])
+                ld_matrix[index_list[:, None], index_list] = temp_ld_S
+                return outlier_likelihood(
+                    index_list, ld_matrix, z, outlier_tau, len(index_list), 1
+                )
+
+            opizer = optimize(ridge_fun, interval=[0, 1], maximum=True)
+            modi_ld = opizer["maximum"] * modi_ld_S + (1 - opizer["maximum"]) * np.diag(
+                np.diag(modi_ld_S)
+            )
+            outlier_likelihood = carmapy.carmapy_c.outlier_Normal_fixed_sigma_marginal
+            test_log_BF = outlier_likelihood(
+                index_list, ld_matrix, z, outlier_tau, len(index_list), 1
+            ) - outlier_likelihood(
+                index_list, modi_ld, z, outlier_tau, len(index_list), 1
+            )
+            test_log_BF = -abs(test_log_BF)
+            print("Outlier BF:", test_log_BF)
+            print("This is xi hat:", opizer)
+        else:
+            pass
+        return filtered_StudyLocus
+
+    @staticmethod
+    def run_finemapping(
+        filtered_LDMatrix: str,
+        filtered_StudyLocus: str,
+        lead_SNP_ID=str,
+        n_sample=int,
+    ):
+        """Runs the SuSiE fine-mapping algorithm from fine-mapping-inf package.
+        Currently unfinished for running in ETL.
+        TODO
+        - Reaplce file names with filtered_LDMatrix and filtered_StudyLocus
+        - Automate getting n_sample and lead_SNP_ID for a locus
+        - Check susie-inf package dependencies don't cause conflicts
+        """
+        n_sample = self.n_sample
+        susieinf_command = f"""python /Users/hn9/Documents/GitHub/fine-mapping-inf/run_fine_mapping.py \
+            --sumstats /Users/hn9/Documents/GitHub/fine-mapping-inf/susieinf/loci/{lead_SNP_ID}_locus_sumstat_with_dentist.txt.gz \
+            --beta-col-name beta \
+            --se-col-name se \
+            --ld-file /Users/hn9/Documents/GitHub/fine-mapping-inf/susieinf/loci/{lead_SNP_ID}_locus_ukbb_ld.txt.gz \
+            --n {n_sample} \
+            --method susieinf \
+            --save-tsv \
+            --eigen-decomp-prefix {lead_SNP_ID}_locus \
+            --output-prefix  {lead_SNP_ID}_locus """
```

Comment:
I have only skimmed through the code, but having hard coded local path would certainly not work. It needs to be somehow parametrized.

Reply:
Yes I agree. Also I suspect, if we are happy to use this susie-inf package, that we can take their functions directly instead - but I’m not sure how this is best implemented in the ETL so I’ve left this command line code as a placeholder for now.

Reply:
Small update on this: I've updated the code so it's setup to try and run the susie functions copy-pasted from the finemapping-inf package directly (tested it [here](https://github.com/hlnicholls/slalom-susie/blob/main/python-pipeline-tests/susie-inf-test/susie-inf-test.ipynb) but not tested running in the ETL itself). I think this will still need a lot of work, but hopefully it's a better starting point